Adapt all scripts to work together with build-packages.sh

redtide opened the pull request:
- Added `build_packages.sh` script, not replacing yet the original `build-packages.sh`.
It redirects to the `build.sh` and a new `package.sh` script used now also for CI. The latter uses the `meson install` step for packaging.
- Reworked the GH Actions script to be more generic, moving most of the bash scripting to the files ~~and enabling automatic releases~~.
- Added a minimal README to describe the scripts.

Closes #423 (InnoSetup packages only) and closes #491.

This requires some patience for testing, so please help.

franko commented:
Do you need some help to test the new script of this is still work-in-progress ?

redtide commented:
Yes, some collaboration would be appreciated.

franko commented:
Ok, I have given a look and I think that the "static" flag that explicitly use lhelper should disappear. The script should be lhelper-agnostic and static/shared library agnostic. It should just let Meson find the library and use it.

redtide commented:
That is how it works, that custom flag is to tell the script if to install and build with lhelper or install the dependencies.

redtide commented:
Looks like that you want only to find problems instead solving them here, if you want just tell me if I need to close this, delete all stuff and never go back to this. Asking help multiple times, being ignored in favor of other issues and then get back after a while pointing out defects after 2 months of my time is not what I call collaboration.

franko commented:
I do not agree, the scope of the build script is to build the application and create the package. To install lhelper, the library or other stuff is not its business. This is the principle of the separation of the concerns.

In addition this is subtly wrong to me: it is not because you are using the -static flag that it has to proceed and install lhelper on my system. Lhelper maybe already installed on my system. In addition lhelper is used the wrong way. The  lhelper environment that contains the libraries is supposed to persist so that you don't always build the libraries over and over again.

I think your view of lhelper iss biased because you only used it in a CI environment where the machine starts from scratch every time. Normally you create the lhelper environment for a specific application with the needed libraries only once. Then you use the same environment over and over again without having to recompile the libraries each time.

Otherwise I quite like your scripts, they are very well organized and complete. Please do this change so that I can proceed with the tests on my side.

redtide commented:
I don't understand why don't you do it since it is your software which I don't understand, I just try to make it work.

redtide commented:
Note that probably I did it that way for convenience because in each `run: bash` a new bash instance is run, so the previous variables are lost. In MSYS2 I need to pass the path of LHelper libs and includes to CFLAGS or it doesn't find SDL, don't ask me why but it works that way on GHA.

```bash
  if [[ "$OSTYPE" == "msys" ]]; then
    CFLAGS=-I${LHELPER_ENV_PREFIX}/include/SDL2
    LDFLAGS=-L${LHELPER_ENV_PREFIX}/lib
  fi
```

So instead run it separately I added that --static parameter in the build because needed to tell the appimage to not package the dynamic libraries and in the package script to keep variables needed (and I don't remember if even active the environment it creates).
Now I'm making the script separated but I expect it will not work so if you can please checkout my branch and make a diff that I can apply later to fix the use of lhelper or just a pastebin code will help a lot.

I have also the problem with the source package I asked in the other discussion:
> do you know if there is a simple way to call `meson setup` to configure only start.lua & friends without checking for the dependencies, without having to add some `configure_only` / `no_dep_check` option?

franko commented:
> Note that probably I did it that way for convenience because in each `run: bash` a new bash instance is run, so the previous variables are lost. In MSYS2 I need to pass the path of LHelper libs and includes to CFLAGS or it doesn't find SDL, don't ask me why but it works that way on GHA.

Normally I work with both MINGW32 and MINGW34 and SDL2 with lhelper works out-of-the-box using the script sdl-config script (which we require in the meson build).

If there is a bug with lhelper please open an issue in lhelper with some details.

I know that CI is painful to deal with when you have a problem because you cannot log in and see what happens but I cannot do anything about that. In sourcehut when you have a CI job you can log in the machine, have a shell and see what happens.

Normally you can use the command `lhelper list files sdl2` from within the environment (use the command "activate" for the environment) to see the files installed for the SDL library. You can look at the lhelper log files and you can run the sdl-config script manually from the environment to see what's wrong.

> So instead run it separately I added that --static parameter in the build because needed to tell the appimage to not package the dynamic libraries and in the package script to keep variables needed (and I don't remember if even active the environment it creates).

I don't know appimage but it theory it should be able to figure out which library is needed or not. In any case I understand the problem but this doesn't mean you go straight and install lhelper each time.

> Now I'm making the script separated but I expect it will not work so if you can please checkout my branch and make a diff that I can apply later to fix the use of lhelper or just a pastebin code will help a lot.
> 
> EDIT: [Here](https://github.com/redtide/lite-xl/actions/runs/1196929669) the build of a tag test.

Sorry, I do not have time to look at that at the moment.

I tested the new build_packages.sh script on windows using MINGW64 with TDM-GCC and an already existing lhelper environment (that one I always use on my PC, I keep the environment). I just used the `--portable` option and nothing else.

The script worked completely fine and the final package was functional but there are some things to fix I think:

1. it doesn't use the meson option `--buildtype=release` so I guess it is doing a debug build
2. it is installing and adding in the ZIP file all the files from subprojects, notably the lua and libagg stuff. Of course this shouldn't be included in the package. To me Meson is a little bit dumb about that but I thought you was given some explicit flags to avoid this.
3. the package is called mingw64 but it shouldn't be named like this. To the final user it doesn't mean anything because it is just an executable that just works on any windows install. We just need to precise the CPU architecture so x86-64 or x86 (i686 ?) or arm or whatever.

Also some side note: to produce a standalone executable I needed to use TDM GCC. Otherwise it depends on some libraries like libgcc.dll libwinpthread.dll and things like this. I may need to know this to produce the final release packages.

Also another small note: can we simplify the option so on Windows it figures out by himself it should be a portable build ? As you already pointed out a unix-like build doesn't make a lot of sense on windows.

redtide commented:
> > Note that probably I did it that way for convenience because in each `run: bash` a new bash instance is run, so the previous variables are lost. In MSYS2 I need to pass the path of LHelper libs and includes to CFLAGS or it doesn't find SDL, don't ask me why but it works that way on GHA.
> 
> Normally I work with both MINGW32 and MINGW34 and SDL2 with lhelper works out-of-the-box using the script sdl-config script (which we require in the meson build).
> 
> If there is a bug with lhelper please open an issue in lhelper with some details.
> 
> I know that CI is painful to deal with when you have a problem because you cannot log in and see what happens but I cannot do anything about that. In sourcehut when you have a CI job you can log in the machine, have a shell and see what happens.

As I said it doesn't work on GHA if you don't tell where the static lib and the includes are, same on my MSYS2 environment on W10.

> Normally you can use the command `lhelper list files sdl2` from within the environment (use the command "activate" for the environment) to see the files installed for the SDL library. You can look at the lhelper log files and you can run the sdl-config script manually from the environment to see what's wrong.
> 
> > So instead run it separately I added that --static parameter in the build because needed to tell the appimage to not package the dynamic libraries and in the package script to keep variables needed (and I don't remember if even active the environment it creates).
> 
> I don't know appimage but it theory it should be able to figure out which library is needed or not. In any case I understand the problem but this doesn't mean you go straight and install lhelper each time.
> 

>     1. it doesn't use the meson option `--buildtype=release` so I guess it is doing a debug build

not even yours when not using PGO, but stripping the exe after build, and so I did in the same way.

>     2. it is installing and adding in the ZIP file all the files from subprojects, notably the lua and libagg stuff. Of course this shouldn't be included in the package. To me Meson is a little bit dumb about that but I thought you was given some explicit flags to avoid this.

There is a step where those files should be removed, I did that workaround because using `--skip-subprojects` requires Meson 0.58, I added a note about it, if this doesn't work then is something happened after some change and I didn't notice it, I will fix this later.

>     3. the package is called mingw64 but it shouldn't be named like this. To the final user it doesn't mean anything because it is just an executable that just works on any windows install. We just need to precise the CPU architecture so x86-64 or x86 (i686 ?) or arm or whatever.

To me makes sense because it specifies the arch anyway, plus explain that is not using MSVC runtimes, but fine, I can change it.

> Also some side note: to produce a standalone executable I needed to use TDM GCC. Otherwise it depends on some libraries like libgcc.dll libwinpthread.dll and things like this. I may need to know this to produce the final release packages.
> 

And you tell me this only now?

> Also another small note: can we simplify the option so on Windows it figures out by himself it should be a portable build ? As you already pointed out a unix-like build doesn't make a lot of sense on windows.

On windows should not be necessary to specify it, it should on other cases. If it's not the case I will check this other detail later too.

Could you please take care of the lhelper part to make work the releases? I've had enough.

franko commented:
> As I said it doesn't work on GHA if you don't tell where the static lib and the includes are, same on my MSYS2 environment on W10.

If there is an issue with lhelper please open an issue in the lhelper project. I cannot fix an issue that is not reported and I never observed myself.

> not even yours when not using PGO, but stripping the exe after build, and so I did in the same way.

This was an error on my side. I cannot believe I was shipping packages compiled in debug mode, too bad.

> There is a step where those files should be removed, I did that workaround because using `--skip-subprojects` requires Meson 0.58, I added a note about it, if this doesn't work then is something happened after some change and I didn't notice it, I will fix this later.

Don't know but FYI, I was using meson 0.58.1 if that helps. I agree that removing the files is a good solution it just has to be done properly. The spurious installed files was inside a folder named "francesco" besides the data directory so I guess there is something wrong.

> To me makes sense because it specifies the arch anyway, plus explain that is not using MSVC runtimes, but fine, I can change it.

This is just an executable, people don't need to know how it was compiled as long as it works correctly and requires no additional runtimes.

> And you tell me this only now?

Yes, it come to my mind this morning when I tried to run the new script. I took me years to perfect the toolchain I use to produce binaries on windows.

I told you you are in charge of the next release so if you create a package that requires MSYS DLLs and you want to package them I am okay with that. I am just telling you that if you don't use TDM GCC you need to bundle some DLLS.

> Could you please take care of the lhelper part to make work the releases? I've had enough.

I just noted that there is a issue you have yet to report on the lhelper project's page.

Please do not complain if you wanted to do the release and now there are some details we need to fix. Please stop complaining about lhelper and stop complaining that if you discard all the script I made and the solution I found you now have to solve the issues I already solved in the past.

Now you added Meson among the variables because you wanted to use meson install so this make for more problems. I was just copying the files I needed for the package, full stop, so I didn't have any issue about Meson what it was doing or not.

You are also having additional issue because we run in the CI environment. I didn't ask for this additional problems because I know it is an hassle.

It is you that wanted automatic packages generated by CI. I told you you are not forced to use lhelper but I wanted static linking and you can ditch lhelper and use whatever you want as long as it is static linking.

Now if you want me drop also this requirement about static linking and you want to ship lite-xl with a bunch of DLL please go ahead, I will let you release the packages like that.

I told you you are in charge to create the packages for the next release and you can do whatever you think is better. I was just trying to help with the new script since you asked to have some help.

redtide commented:
LHelper stuff is now in its own script, so all I asked is to make it work on your side, and if it does try to see what problem could be on CI.
I don't even know what to report, it was working like that and producing the files previously and it was OK to me, it's not a bug to me to pass some CFLAGS. It just doesn't work anymore by not finding the sdl library, so it needs to be tested, that's all.

It might be my fault, IDK, I'm just tired to test on something I don't understand and I don't want to back again to make dll based build releases or build static libs myself changing things another time.
If you can help OK, if not OK anyway.

redtide commented:
@franko I think to disable again the releases and if there is no other requirements replace also your current `build-packages.sh` to merge this and fix releases later, is it OK?

redtide commented:
I forgot: I guess if we use the release buildtype in meson the strip command should go, no?

franko commented:
> @franko I think to disable again the releases and if there is no other requirements fix the above you mentioned, replace also your current `build-packages.sh` to merge this and fix releases later on another commit/PR, is it OK?

I am okay to go. About the problem you got with lhelper on MINGW32 you can add the flag with the SDL2 include directory unconditionally or may be just for Windows. In the case it is not needed it will make no harm because it is just an include directory we need to include anyway.

So that should be something like:

```sh
if [ we are on windows ]; then
  CFLAGS+="$(lhelper dir)/include/SDL2"
fi
```

The command `lhelper dir` just return the prefix directory for the environment currently loaded.

franko commented:
> I forgot: I guess if we use the release buildtype in meson the strip command should go, no?

The strip should stay, please. The release build and the strip do different things but I guess you know this.

In the meson website https://mesonbuild.com/Builtin-options.html we have an option to strip the executable so may may use it as well.

franko commented:
When you are ready to go with this I will make a few more tests and I will merge if everything is fine.

Let's say that if you do in this PR the commit to replace the old script with the new one it means we are ready to go.

redtide commented:
Thanks, say that you will see it green instead grey when done.

franko commented:
> Note that probably I did it that way for convenience because in each `run: bash` a new bash instance is run, so the previous variables are lost. In MSYS2 I need to pass the path of LHelper libs and includes to CFLAGS or it doesn't find SDL, don't ask me why but it works that way on GHA.
> 
> ```shell
>   if [[ "$OSTYPE" == "msys" ]]; then
>     CFLAGS=-I${LHELPER_ENV_PREFIX}/include/SDL2
>     LDFLAGS=-L${LHELPER_ENV_PREFIX}/lib
>   fi
> ```
> 
> So instead run it separately I added that --static parameter in the build because needed to tell the appimage to not package the dynamic libraries and in the package script to keep variables needed (and I don't remember if even active the environment it creates).
> Now I'm making the script separated but I expect it will not work so if you can please checkout my branch and make a diff that I can apply later to fix the use of lhelper or just a pastebin code will help a lot.
> 
> EDIT: [Here](https://github.com/redtide/lite-xl/actions/runs/1196929669) the build of a tag test.

I tried to install sdl2 using lhelper in MINGW32 and everything was working. I tried to look at the CI run you was pointing out but I didn't understood clearly what was going on. I have seen there are no log files. If you can point out to me what went wrong I may have a better look.

redtide commented:
IDK what cause the problem, I only suspected 2 things: a recent change of yours on lhelper master that CI didn't like (I have not checked if you did, so just a guess) or something changed on the CI itself. I need some patience later to back to it, for now all I want is to see what is missing to fix the normal builds.
The only thing I know is that on each `run: bash` line in the YML file a new bash instance is run, so might be a lost setting on a previous one.

franko commented:
> The only thing I know is that on each `run: bash` line in the YML file a new bash instance is run, so might be a lost setting on a previous one.

If you are starting a new bash you need just to make the call `source "$(lhelper env-source lite-xl)"` or whatever the environment name is instead of "lite-xl". It will "activate" the environment in the current shell process. For activate I mean it will configure some environment variables like PATH, PKG_CONFIG_PATH etc, all is needed for the environment.

redtide commented:
OK, I think I know what to do for the release builds but I'll need some time, so delay it for a later change.
For the rest should be good to go, please check.

franko commented:
That looks all fine now, thank you.

I tested on windows using mingw64 and my already existing lhelper environment and all works fine.

The final executable only depend on an additional DLL that is `libwinpthread-1.dll`. I copied the file from /mingw64/bin beside the executable and it was working. Otherwise please note that the package is not functional, when you double click in the executable windows will tell you that the libwinpthread dll is missing.

We have two possible solutions for this problem. Either we copy the missing dll into the package beside the lite-xl executable either we swtich to TDM GCC. I advice for the moment to go for the first solution that should be easily integrated in your script.

I future we may look to use TDM GCC but in terms of automated build this is more complicated so for the moment we can leave this beside.

In the future I thought also about adding back the option to make a pgo build in case the package is created manually by a maintainer.

Thank you again for this work, I appreciate your patience and I apologize for being harsh sometimes but steering an open source project with many collaborators is a little bit difficult for me but I am learning.

redtide commented:
I see, I'm not against TDM GCC if it will work properly on CI, the PGO can be added back on buiild-packages,sh, I don't use it at all for CI.
Now I find out there is also a problem on macOS I haven't noticed, so I think I'll do some fix to add the missing libs for all builds and fix this problem; for some reason it creates the bundle dir instead the normal one when not passing the bundle option, so the "unix like" package is not built correctly. 😞

franko commented:
The PGO is already back in the master branch, I did it in five minutes. I like how your scripts are well structured even if a little bit verbose.

By the way, on windows with MINGW64 PGO produce an executable that crashes but it works okay with MINGW32. PGO also doesn't work on macOS with clang. Just in case you wanted to try FYI.

As for the problems we need to fix, no need to hurry, we have all the time we need before the next release.

redtide commented:
OK, I left a comment on your commit, it seems the change stalled the build.

franko commented:
Even if I don't like these kind of expressions what I am using is idiomatic bash AFAICT, see for example:

https://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash

I think the problem is that the process is running with the venerable bash version shipping with macOS, version 3.2.57. Normally on macOS we should ensure a recent version of bash is used as provided for example by homebrew. I strongly advice you avoid using the old bash version as we may stumble in strange problems like this.

Otherwise we can simplify, if we don't want to make the difference between a variable that is unset and one that is set to an empty string we can use: `if [ -n "$pgo" ]`, it will be fine in our case.

redtide commented:
No, I think I'll force to update bash in any case, not only when using lhelper and leave as you did. But at this point I realize that my scripts are working only on bash because double square brackets (and IDK if else). I'm going to push a fix directly also for the previously mentioned package issue on mac. What do you think?

franko commented:
Ok for me, please proceed directly with the changes you need.

Let's not pretend our script can work with any shell: we need a recent bash version to simplify our lives but having a recent bash version seems a very reasonable requirement.

redtide commented:
#496, can't push on upstream.
